iscservice/demo/iscdemo.c

```diff
@@ -0,0 +1,607 @@
+	buf = malloc(x->mr_rbuf.b_nob + y->mr_lbuf.b_nob + 1);
+	if (buf == NULL) {
+		fprintf(stderr, "failed to allocate %d of memory for result\n",
+			                            len);
+		return NULL;
+	}
+
+	/*
+	 * Glue two edge buffers from the two results and read the value
+	 * from it. This value must be accounted in the final computation.
+	 */
+	memcpy(buf, x->mr_rbuf.b_addr, x->mr_rbuf.b_nob);
+	buf[x->mr_rbuf.b_nob] = '\0';
+	DBG2("xrbuf=%s\n", buf);
+	memcpy(buf + x->mr_rbuf.b_nob, y->mr_lbuf.b_addr, y->mr_lbuf.b_nob);
+	buf[x->mr_rbuf.b_nob + y->mr_lbuf.b_nob] = '\0';
+
+	rc = sscanf(buf, "%lf%n", &val1, &len);
+	if (rc < 1) {
+		fprintf(stderr, "failed to read the resulting xr-value\n");
+		m0_free(buf);
+		return NULL;
+	}
+
+	y->mr_idx += x->mr_nr;
+	y->mr_nr  += x->mr_nr;
+
+	DBG2("buf=%s val1=%lf\n", buf, val1);
+
+	/*
+	 * It may happen that there will be two values in the buffer (when
+	 * the split happens right around the delimiter), so we should try
+	 * to read it and account it in the final computation too.
+	 */
+	val2 = val1;
+	rc = sscanf(buf + len, "%lf", &val2);
+	if (rc == 1) {
+		DBG2("val2=%lf\n", val2);
+		y->mr_idx++;
+		y->mr_nr++;
+	}
+
+	DBG2("xval=%lf yval=%lf\n", x->mr_val, y->mr_val);
+
+	if (ICT_MIN == op_type)
+		y->mr_val = min3(min_check(x->mr_val, y->mr_val), val1, val2);
+	else
+		y->mr_val = max3(max_check(x->mr_val, y->mr_val), val1, val2);
+
+	/* Update the resulting value index. */
+	if (y->mr_val == x->mr_val)
+		y->mr_idx = x->mr_idx;
+	else if (y->mr_val == val1)
+		y->mr_idx = x->mr_nr;
+	else if (y->mr_val == val2)
+		y->mr_idx = x->mr_nr + 1;
+
+	m0_free(buf);
+
+	return y;
+}
+
+enum elm_order {
+	ELM_FIRST,
+	ELM_LAST
+};
+
+static void set_idx(struct mm_result *res, enum elm_order e)
+{
+	if (ELM_FIRST == e)
+		res->mr_idx = 0;
+	else
+		res->mr_idx = res->mr_nr;
+}
+
+/**
+ * Process the first or the last values in the first or last results.
+ */
+static void check_edge_val(struct mm_result *res, enum elm_order e,
+			   enum isc_comp_type type)
+{
+	const char *buf;
+	double      val;
+
+	if (ELM_FIRST == e)
+		buf = res->mr_lbuf.b_addr;
+	else // last
+		buf = res->mr_rbuf.b_addr;
+
+	if (sscanf(buf, "%lf", &val) < 1) {
+		fprintf(stderr, "failed to parse egde value=%s\n", buf);
+		return;
+	}
+
+	DBG2("edge val=%lf (%s)\n", val, (ELM_FIRST == e) ? "first" : "last");
+
+	if (res->mr_nr == 1) {
+		res->mr_val = val;
+		return;
+	}
+
+	if (ICT_MIN == type && val < res->mr_val) {
+		res->mr_val = val;
+		set_idx(res, e);
+	} else if (ICT_MAX == type && val > res->mr_val) {
+		res->mr_val = val;
+		set_idx(res, e);
+	}
+}
+
+static void mm_result_free_xcode_bufs(struct mm_result *r)
+{
+	m0_free(r->mr_lbuf.b_addr);
+	m0_free(r->mr_rbuf.b_addr);
+}
+
+static void *minmax_output_prepare(struct m0_buf *result,
+				   bool last_unit,
+				   struct mm_result *prev,
+				   enum isc_comp_type type)
+{
+	int               rc;
+	struct mm_result  new = {};
```

Comment:
I am sorry, I still don't understand the "struct mm_result".
The "mm_result" is the result from one request, (i.e. from a single unit).
It contains the min/max value and the number of items in this unit.
The final result should be aggregated from all the requests.
For min, we need to compare all the results from all requests, and find the minimal value from them.

But why do we need something called "edge"?
some cutting value?

This makes things complex. we can just suppose the content of a unit is an array of "double float" number, with size of 8 bytes. So, we just need to search the array and find the minimal value.
When all requests returned, we just need to aggregate the results.

Please explain your algorithm with details.
As a demo, we need something very simple and make it easy to understand.

Reply:
Now, I understand the reason.
We are treating the content of the object/unit, as human-readable double float number.
So, we need to use fscanf()/sscanf() to read the number.

So, I still have the question.
Why don't we treat the object content as array of 64-bit double float number?
This is more common in data format for scientific data.

Reply:
The purpose of the demo was to demonstrate the in-storage computation usage in general case, not only the case when the data records are aligned to the units boundaries (which is over-simplistic in general case). And how to handle such issues when the data records may be split between units.

The purpose was agreed with @nikitadanilov from the beginning of the demo improvement work.